libnmstate/nm/route.py

```diff
@@ -0,0 +1,110 @@
+    for active_connection in client.get_active_connections():
+        ip_cfg = func['get_ip_cfg_func'](active_connection)
+        if not ip_cfg:
+            continue
+
+        iface_name = None
+        # If we have master device, we use it, else we use first device.
+        if hasattr(active_connection, 'master'):
+            iface_name = active_connection.master.get_iface()
+        else:
+            devs = active_connection.get_devices()
+            if devs:
+                iface_name = devs[0].get_iface()
+        if not iface_name:
+            logging.warning(
+                'Got connection {} has not interface name'.format(
+                    active_connection.get_id()))
+            continue
+
+        skip_networks.extend(
+            list('{}/{}'.format(addr.get_address(), addr.get_prefix())
+                 for addr in ip_cfg.get_addresses()))
+        static_routes = []
+        ip_profile = func['get_ip_profile_func'](active_connection)
+        if ip_profile:
+            for route in ip_profile.props.routes:
+                static_routes.append(
+                    '{ip}/{prefix}'.format(
+                        ip=route.get_dest(), prefix=route.get_prefix()))
+        table_id = _get_route_table_id(active_connection, ip_cfg,
+                                       func['get_ip_profile_func'])
+        for route in ip_cfg.get_routes():
+            dst = '{ip}/{prefix}'.format(
+                ip=route.get_dest(), prefix=route.get_prefix())
+            if _should_skip(dst, skip_networks):
+                continue
+            route_origin = _get_route_origin(
+                active_connection,
+                static_routes,
+                dst,
+                func['is_auto_route_func'],
+                auto_origin_value)
+            next_hop = route.get_next_hop()
+            if not next_hop:
+                next_hop = ''
+            metric = int(route.get_metric())
+
+            route_entry = {
+                'origin': route_origin,
+                'table-id': table_id,
+                'table-name': iplib.get_route_table_name(table_id),
+                'destination': dst,
+                'next-hop-interface': iface_name,
+                'next-hop-address': next_hop,
+                'metric': metric,
+            }
+            ret.append(route_entry)
+    ret.sort(key=itemgetter('table-id', 'destination'))
+    return ret
+
+
+def _should_skip(dst, skips):
+    '''
+    Check whether specified route destination is in skip networks.
+    '''
+    for network in skips:
+        if iplib.is_subnet_of(dst, network):
+            return True
+    return False
+
+
+def _get_route_origin(active_connection, static_routes, dst,
+                      is_auto_route_func, auto_origin_value):
+    if dst not in static_routes and is_auto_route_func(active_connection):
+        return auto_origin_value
+    return 'static'
+
+
+def _get_route_table_id(active_connection, ip_cfg, get_ip_profile_func):
```

Comment:
IMHO you could just pass the ip_profile instead of the function as an argument. Also the ip_cfg argument does not seemt to be used.

Reply:
Done. Thanks.